src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -235,6 +262,74 @@ internal static void SetProperty(object resource, string propertyName, object va
             }
         }
 
+        internal static void SetInstanceAnnotations(ODataResource oDataResource, IODataInstanceAnnotationContainer instanceAnnotationContainer, 
+            ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
+        {
+            if(oDataResource.InstanceAnnotations != null && oDataResource.InstanceAnnotations.Count > 0)
```

Comment:
> && oDataResource.InstanceAnnotations.Count > 0 [](start = 56, length = 47)

You should no longer need this check; if there are no instance annotations, your foreach will just exit on first comparison.